sklearn/naive_bayes.py

```diff
@@ -958,3 +985,252 @@ def _joint_log_likelihood(self, X):
+
+        def _unique_sums(X_feature_class, Y_class):
+            # Sum for each category in x the corresponding values in y
+            class_cats, inv = np.unique(X_feature_class, return_inverse=True)
+            n_feature_class = np.zeros(len(class_cats), dtype=Y_class.dtype)
+            np.add.at(n_feature_class, inv, Y_class)
+            return class_cats, n_feature_class
+
+        self.class_count_ += Y.sum(axis=0)
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            cats = np.unique(X_feature)
+            _update_cat_mapping(self.feature_cat_index_mapping_[i], cats)
+            # update category_count_dimensions in case partial_fit is used, to
+            # to account for unseen categories
+            self.category_count_[i] = _update_cat_count_dims(
+                self.category_count_[i], self.feature_cat_index_mapping_[i])
+            _update_cat_count(X_feature, Y,
+                              self.category_count_[i],
+                              self.feature_cat_index_mapping_[i],
+                              self.class_count_.shape[0])
+
+    def _update_feature_log_prob(self, alpha):
+        feature_log_prob = {}
+        for i in range(self.n_features_):
+            smoothed_cat_count = self.category_count_[i] + alpha
+            smoothed_class_count = smoothed_cat_count.sum(axis=1)
+            feature_log_prob[i] = (np.log(smoothed_cat_count) -
+                                   np.log(smoothed_class_count.reshape(-1, 1)))
+        self.feature_log_prob_ = feature_log_prob
+
+    def _joint_log_likelihood(self, X):
+        def _get_cat_index(mapping, cat, handle_unknown):
+            try:
+                return mapping[cat]
+            except KeyError:
+                if handle_unknown == 'warn':
+                    warnings.warn(
+                        "Category {} not expected for feature {} "
+                        "of features 0 - {}."
+                        .format(cat, i, self.n_features_-1)
+                    )
+                elif handle_unknown == 'error':
+                    raise KeyError(
+                        "Category {} not expected for feature {} "
+                        "of features 0 - {}."
+                        .format(cat, i, self.n_features_-1)
+                    )
+                else:
+                    return None
+
+        if not X.shape[1] == self.n_features_:
+            raise ValueError("Expected input with %d features, got %d instead"
+                             .format(self.n_features_, X.shape[1]))
+        jll = np.zeros((X.shape[0], self.class_count_.shape[0]))
+        for i in range(self.n_features_):
+            mapping = self.feature_cat_index_mapping_[i]
+            # generator over all samples of feature i
+            indices = (_get_cat_index(mapping, cat, self.handle_unknown)
+                       for cat in X[:, i])
+            indices = [idx for idx in indices if idx is not None]
```

Comment:
you're still manually performing a `O(n_features * n_samples)` loop here. I'm concerned about efficiency.

Reply:
your solution? @NicolasHug list of np.array instead of list of dictionary?
if we assume that n_features << n_samples , maybe O(n_features * n_samples) is acceptable?

Reply:
Thinking on the subway, maybe list of np.array is indeed more natural and much faster?

Reply:
Even if we go for `np.array`, at some point we have to map somehow all the categories to indices. This could be done in a preprocessing step but it needs to be done and thus we will have O(n_features * n_samples) which btw is not problematic if the operation is fast. In general, I would say that all estimators in Scikit-Learn have at least O(n_features * n_samples) since otherwise it would mean that certain features or samples are not used at all, right? The only difference is that in most cases the n_samples access is vectorized by using NumPy. If we allow the categories to be anything like a string or like a huge number, this operation cannot be efficiently vectorized with NumPy as far is I understand. The problem is the mapping to an index which always depends on how many categories we have seen so far, thus it's not parallelizable. We could use a trick like hashing or so as several Big Data frameworks do but this would not work well together with NumPy...

As you see @timbicker and I are a little lost here. So we would appreciate some more pointers on how to solve this problem. Thanks.